File: solutions/javascript/matching-brackets/1/matching-brackets.js
//
// This is only a SKELETON file for the 'Matching Brackets' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const isPaired = (input) => {
  const stack = [];
  const pairs = {'(': ')', '[': ']', '{': '}'};

  for (const char of input) {
    if ("([{".includes(char)) {
      stack.push(pairs[char]);  // push expected closing bracket
    } else if (")]}".includes(char)) {
      if (stack.pop() !== char) {
        return false;
      }
    }
  }

  return stack.length === 0;
};

File: solutions/python/matching-brackets/1/matching_brackets.py
def is_paired(input_string):

    stack = []

    pairs = {
        ')': '(',
        ']': '[',
        '}': '{'
    }

    for ch in input_string:

        
        if ch in "([{":
            stack.append(ch)

        
        elif ch in ")]}":

            if not stack:
                return False

            if stack[-1] != pairs[ch]:
                return False

            stack.pop()

    return len(stack) == 0

